### PR TITLE
Allow Union of scalars to be NumberType

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68773
* #68772
* #66592
* __->__ #66591

Differential Revision: [D31632599](https://our.internmc.facebook.com/intern/diff/D31632599)

### CODE REVIEW

**Review by gmagogsfm — COMMENTED** [review]
Could you add a test? Thanks

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632599).

File: test/test_jit.py [hunk 1]
```diff
@@ -11125,6 +11125,12 @@ def fn(x, y):
 
         self.checkScript(fn, (torch.randn(3, 2, dtype=torch.float), torch.ones(3, 2, dtype=torch.float)))
 
+    def test_union_to_number(self):
+        @torch.jit.script
+        def fn(x: Union[int, complex, float], y: Union[int, complex, float]):
+            return x + y
+        FileCheck().check(": Scalar):").run(fn.graph)
```
[the remark this answers is not in the document]
Comment (tugsbayasgalan) [reply test/test_jit.py:11132]: https://github.com/pytorch/pytorch/blob/e2be087207c210a30e201b786fc4d3aab8336837/torch/csrc/jit/frontend/schema_type_parser.cpp#L57 (This is how its represented in the frontend I think)

File: torch/jit/annotations.py [hunk 2]
```diff
@@ -411,6 +414,7 @@ def ann_to_type(ann, loc):
     'TensorType',
     'TupleType',
     'FloatType',
+    'NumberType',
```
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply torch/jit/annotations.py:417]: If it is only about a possible future use case, then let's not add it now. 

Fun read about YAGNI:
https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

**Review by gmagogsfm — APPROVED** [review]

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632599).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632599).

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31632599).

**Pull request closed without merging**